ads/inabox/inabox-messaging-host.js

```diff
@@ -209,35 +209,111 @@ export class InaboxMessagingHost {
+   *       hierarchy that the outer host window can still measure, which is
+   *       the top most cross domained frame, or the creative frame.
+   * EXAMPLE:
+   *   If we have a frame hierarchy:
+   *     Host -> Friendly Frame -> X Domain Frame 1 -> Message Source Frame
+   *     and "Friendly Frame" is registered in this.iframes_, then
+   *     "Message Source Frame" is valid, because one of its parent frames
+   *     is registered in this.iframes_, and the result of the call to
+   *     getFrameElement_ would be the iframe "X Domain Frame 1" as it is
+   *     the deepest frame that the host doc can accurately measure.
+   * Note: The sentinel should be unique to the source window, and the result
    * is cached using the sentinel as the key.
    *
    * @param source {!Window}
    * @param sentinel {string}
-   * @returns {?HTMLIFrameElement}
+   * @return {?Element}
    * @private
    */
   getFrameElement_(source, sentinel) {
     if (this.iframeMap_[sentinel]) {
       return this.iframeMap_[sentinel];
     }
-
-    // Walk up on the window tree until find host's direct child window
-    while (source.parent !== this.win_ && source !== this.win_.top) {
-      source = source.parent;
-    }
-
-    // Find the iframe element that hosts the message source window
+    const measureableFrame = this.getMeasureableFrame(source);
+    const measureableWin = measureableFrame.contentWindow;
     for (let i = 0; i < this.iframes_.length; i++) {
       const iframe = this.iframes_[i];
-      if (iframe.contentWindow === source) {
-        // Cache the found iframe with its sentinel.
-        this.iframeMap_[sentinel] = iframe;
-        return iframe;
+      for (let j = 0, tempWin = measureableWin;
+        j < 10; j++, tempWin = tempWin.parent) {
+        if (iframe.contentWindow == tempWin) {
+          this.iframeMap_[sentinel] = measureableFrame;
+          return measureableFrame;
+        }
+        if (tempWin == window.top) {
+          break;
+        }
       }
     }
     return null;
   }
+
+  /**
+   * Returns whichever window in win's parent hierarchy is the deepest window
+   * that is measurable from the perspective of the current window.
+   * For when win is nested within a x-domain frame, walks up the window's
+   * parent hierarchy until the top-most x-domain frame in the hierarchy
+   * is found. Then, it returns the frame element for that window.
+   * For when win is friendly framed, returns the frame element for win.
+   * @param {!Window} win
+   * @return {?Element}
+   * @visibleForTesting
+   */
+  getMeasureableFrame(win) {
+    // First, we try to find the top-most x-domain window in win's parent
+    // hierarchy. If win is not nested within x-domain framing, then
+    // this loop breaks immediately.
+    let topXDomainWin;
+    for (let j = 0, tempWin = win; j < 10 && tempWin != tempWin.top;
+      j++, tempWin = tempWin.parent) {
+      if (canInspectWindow(tempWin)) {
```

Comment:
Can we not just move canInspectWindow into the for loop declaration:

 for (let j = 0, tempWin = win; j < 10 && tempWin != tempWin.top && !canInspectWindow(tempWin); topXDomainWin = tempWin, j++, tempWin = tempWin.parent);